### PR TITLE
build: Modifying cmake file(s) to allow for easier cross compiling

### PR DESCRIPTION
**Pull request by jdavidberger**
Cross compiling using a CMAKE_TOOLCHAIN file somewhat breaks the cmake build since it tries to run a handful of compiled test programs. 

This change checks to see if it is cross compiling, and if so simply omits the tested features which require a program to run. This mostly seems to just mean that it falls back to 'strerror' instead of any of the strerror_r variants, and that it uses select instead of polls. This should not prevent whomever is compiling from manually configuring those parameters should they require it.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -785,7 +785,14 @@ endif(HAVE_SIGNAL_FUNC AND HAVE_SIGNAL_MACRO)
 check_symbol_exists(uname          "${CURL_INCLUDES}" HAVE_UNAME)
 check_symbol_exists(strtoll        "${CURL_INCLUDES}" HAVE_STRTOLL)
 check_symbol_exists(_strtoi64      "${CURL_INCLUDES}" HAVE__STRTOI64)
-check_symbol_exists(strerror_r     "${CURL_INCLUDES}" HAVE_STRERROR_R)
+
+if(NOT ${CMAKE_CROSSCOMPILING})
+  # We can't tell which version of strerror_r we have unless we can run
+  # a test program; so if we are cross compiling just assume we don't have
+  # strerror_r. 
+  check_symbol_exists(strerror_r     "${CURL_INCLUDES}" HAVE_STRERROR_R)
```
Comment (Lekensteyn) [inline CMakeLists.txt]: `check_symbol_exists` uses `try_compile` under the hood which [appears to support][1], shouldn't that  be able to detect the availability of this function?

 [1]: http://www.vtk.org/Wiki/CMake_Cross_Compiling#System_introspection

File: CMake/OtherTests.cmake [hunk 2]
```diff
@@ -177,19 +177,20 @@ int main(void) {
...
+    set(CMAKE_REQUIRED_FLAGS "-DHAVE_SYS_POLL_H")
+  endif(HAVE_SYS_POLL_H)
+  check_c_source_runs("
   #ifdef HAVE_SYS_POLL_H
   #  include <sys/poll.h>
   #endif
   int main(void) {
     return poll((void *)0, 0, 10 /*ms*/);
   }" HAVE_POLL_FINE)
+endif()
```
Comment (Lekensteyn) [inline CMake/OtherTests.cmake]: Don't know the purpose of this check, but wouldn't it be sufficient to check whether it compiles?

File: CMakeLists.txt [hunk 3]
```diff
@@ -868,12 +875,15 @@ endforeach(CURL_TEST)
...
-    )
-  curl_internal_test_run(${CURL_TEST})
-endforeach(CURL_TEST)
+
+if(NOT ${CMAKE_CROSSCOMPILING})
+  foreach(CURL_TEST
+      HAVE_GLIBC_STRERROR_R
+      HAVE_POSIX_STRERROR_R
+      )
+    curl_internal_test_run(${CURL_TEST})
```
Comment (Lekensteyn) [inline CMakeLists.txt]: What about modifying the macro definition which then prints a warning and returns?

File: CMakeLists.txt (hunk 1 above)
Comment (jdavidberger) [inline CMakeLists.txt]: Yes, it finds the availability of the function but curl won't compile if you set HAVE_STRERROR_R but don't choose which form it is -- HAVE_GLIBC_STRERROR_R or HAVE_POSIX_STRERROR_R, which does require the try_run check. Possibly I should move this chunk down to the logic for those so it makes a bit more sense?

File: CMake/OtherTests.cmake (hunk 2 above)
Comment (jdavidberger) [inline CMake/OtherTests.cmake]: I'm honestly not 100% sure on this either; but it seemed like the attempt at running was a deliberate choice so I kept it. 

My assumption is that its testing for some particular form of poll and not just for the existence of poll itself. [It seems like non-zero return codes will fail the test](https://github.com/Kitware/CMake/blob/master/Modules/CheckCSourceRuns.cmake) even though they run, so I chalked it up to some platform somewhere not having a form of poll which works with the code.

File: CMakeLists.txt (hunk 3 above)
Comment (jdavidberger) [inline CMakeLists.txt]: That does seem better long-term. I'll make the change.

**Comment by bagder** [comment]
Ping @billhoffman, @Sukender or @Lekensteyn, can I ask you for comments in regards to merging this PR or if it needs more work/comments ?

File: CMakeLists.txt (hunk 1 above)
Comment (Lekensteyn) [inline CMakeLists.txt]: I see, GNU takes a `char*` return type and the other standard `int`. I'd suggest to move it down.

**Comment by Lekensteyn** [comment]
Needs more work, please address that strerror comment and replace `if(${CMAKE_CROSSCOMPILING})` by `if(CMAKE_CROSSCOMPILING` (dropping `${` and `}`) as that is the recommended style.

For future reference:
If docs: https://cmake.org/cmake/help/v3.0/command/if.html
`CMAKE_CROSSCOMPILING` is documented at https://cmake.org/cmake/help/v3.0/variable/CMAKE_CROSSCOMPILING.html#variable:CMAKE_CROSSCOMPILING

**Comment by bagder** [comment]
Stalled. Feel free to re-open if interest is regained.

**Comment by patrikhson** [comment]
Hi, I was looking for some stuff on the web and found this, you're gonna like it, just take a look here <http://sycamopri.zijasa.com/xynphpro>

See you soon, paf@frobbit.se

**Pull request closed without merging**